### PR TITLE
feat: stable parameter hashing. Fixes #69

### LINKED ISSUE #69: improve folder naming scheme
An acknowledged limitation of the current tag (0.1.0-rc3) is that the naming scheme used for specifying the chain of intermediate outputs in snakemake is somehow brittle.

A consequence of this is what we're internally calling "parameter instability": that means that the same ID will be reused if different runs alter the set (or the order) of parameters passed to an invocation. Do note that I use "parameters" here in the sense that they're used in the benchmark yaml: arguments passed to any module entrypoint (and not as method-specific parameterizations).

The problem is in the serialization of the DAG nodes, but it stems from benchmark/dag, in the sense that there the parameter ID is taken from an enumeration (that is therefore context and ordering dependent).

To overcome this, we'd like to explore the use of a Content-Adressable Storage schema of sorts. It's not a proper CAS, since we're not uniquely hashing contents of any file, but rather representing the parameters object.

There is another implicit requirement, and is that the system aims to maintain some form of human-readable identifiers. This is to facilitate browsing of the outputs tree by a human.

To address this, I proposed to use a convention similar to how device aliasig works in linux (i.e., symlinks under ` /dev/disk/by-*`), pointing to a canonical path.

# Limitations

Some limitations have been pointed out by team members. Notably:

- There's a hard limit on a folder name (usually 255 chars).
- There's a limitation on the max path len (4096 chars).
- Housekeeping - even if we get stable hashing, the filesystem will be polluted by folders and links while the user is changing parameter invocations. 

I think the first 2 constrains are still workable, in the sense that it's up to the benchmark designer to either use a sensible naming scheme, or we can always fall back to hashes.

For the last one, I wouldn't even bother, as long as it's documented. True, we can diff the previous run invocation and do some cleanup of the stale (orphaned) folders, but I think it's not worth the hassle. Perhaps we could provide a partial cleanup subcommand, but I think the benchmarked needs to live with that.

For very complex benchmarks, I would argue we need a more scalable solution that is not based on the filesystem, and I would relax the browseability requirement. That can still be achieved by maintaining a mapping between parameter hashes and the parameter dict, but it can be done by a cli utility module that keeps state in a local database.

### PR DESCRIPTION
**Pull request by DanInci**
## Ticket

https://github.com/omnibenchmark/omnibenchmark/issues/69

## Description

This PR tries to address 3 issues:
1. Better modeling of the Parameters, using a dedicated class
2. Parameter instability problem. See #69 
3. Human-friendly browsability of the results for parameters using symlinks

Result looks like this. Hashed folder names are still used by snakemake to built the required hierarchy. However, the hashed folders are hidden from the usual folder explorers, and symlinks are used instead.
Screenshot 2025-04-27 at 20 01 53

Additionally, the parameters dict is now a .tsv file, that contains relevant information.
```tsv
id	base_path	alias_path	parameters
dc8f7f3b627f294cfaf4d3915f8445d24d19512e931ff330ffaf1c4ce28a8294	out/data/D2/default/methods/M2/.dc8f7f3b627f294cfaf4d3915f8445d24d19512e931ff330ffaf1c4ce28a8294	out/data/D2/default/methods/M2/d1-True_e-2	{"d1": true, "e": "2"}
3297cc0b9f48521ab602a4a90143602c416f1f5029c70182a62e6092166d3bc9	out/data/D2/default/methods/M2/.3297cc0b9f48521ab602a4a90143602c416f1f5029c70182a62e6092166d3bc9	out/data/D2/default/methods/M2/d1-True_e-1	{"d1": true, "e": "1"}
```
## Checklist:
- [x] My code follows the code style of this project.
- [x] My CLI method respects the signature defined in the task.
- [x] I have documented the CLI method accordingly.
- [x] I have added tests to cover my changes.
- [x] I have added a CHANGELOG entry.

### CODE REVIEW

**Comment by btraven00** [comment]
Related: #69

**Comment by imallona** [comment]
Tested it locally and works great, thanks!

Random thought. With the hashing now it is possible to run multiple benchmarks adding new folders to the `out` folder without risking overwrites/collisions. Is this something intended? Or, in other words, should we have a check or a warning indicating something along these lines:

> warning: `out` folder is not empty and its layout is consistent with the benchmark yaml

or 

> (hard): warning `out` folder is not empty and its layout is not compatible with the benchmark yaml

?

**Comment by DanInci** [comment]
> Tested it and looks great!
> 
> Another random thought: do we have a testing yaml with convoluted parameters names so the hash and not the human-friendly folder names is kept for these parameters? AFAIK this was part of the spec, but couldn't test it

There are unit tests for the hashing and symlink creation that test out exactly this. In `test_params.py` from like 125 onwards. They might not be extensive enough, but they test some basic corner case scenarios.

**Comment by btraven00** [comment]
For reusing the output folder, I think what I would do is to refuse to overwrite an existing dir, unless a explicit flag is passed.

Multiple runs: I think it's easier to have them in independent output dirs - possibly qualified by some metadata (like a timestamp: `--timestamp-output' or something.

File: omni/benchmark/symlinks.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+        # If too long, append a short 8 characters hash
+        if len(name) > max_len:
+            short_hash = params.hash()[:8]
+            name = name[: max_len - 9] + "_" + short_hash
+        return name
+
+    def _write_params(self, params, hash_path, human_path):
+        params_hash = params.hash()
+        params_file = hash_path / "parameters.json"
```
[the remark this answers is not in the document]
Comment (btraven00) [reply omni/benchmark/symlinks.py:81]: Thinking a bit more, I'm not sure how do we want to evolve this (per-folder params metadata). If we're gonna use it for validation, I'd try to stick to json because of typing / nestedness. If we envision to drop this metadata at the folder level, then probably tsv is fine.

File: tests/software/test_run_with_software.py [hunk 2]
```diff
@@ -105,3 +108,206 @@ def test_run_benchmark_with_software_apptainer():
         _cleanup_snakemake()
     else:
         raise FileNotFoundError(f"Benchmark file {benchmark_file} does not exist.")
+
+
```
Comment (btraven00) [inline tests/software/test_run_with_software.py:112]: I think my only doubt here is whether we can strategize a bit more on tests for orthogonality. For instance, in this case, maybe do not depend on the conda environment for testing parameters, but try and come with a python-only approach (i.e. one that can run in a simple virtualenv, with no system dependencies).

**Review by btraven00 — COMMENTED** [review]
Thanks for the integration! Overall it looks quite good (and robust) to me.
I'm not sure about the comments I made, maybe we can address them after merging.

About merging order, I would have to see if rebasing this on top of the tests refactor is going to be less cumbersome than the other way around.

**Comment by btraven00** [comment]
:whale: to merge when tests pass -  but please let's do a rebase + squash for a cleaner merge :)
